### DISCUSSION TITLE
Canceling tasks

### DISCUSSION THREAD

**Discussion by kpodsiad** [monix/monix#1482] [Q&A]
I was playing with canceling tasks and I found a behavior which I struggle to understand. Given following code:
```scala
object Main extends App {
  implicit val scheduler = Scheduler.fixedPool("my-pool", 5)

  val heavyTask = Task {
    Thread.sleep(2000) // mimic heavy computation
    println("heavy computation done")
  }
    .doOnCancel(Task(println("Task was canceled")))
    //.executeAsync

  val cancelable = heavyTask.runToFuture
  cancelable.cancel

  Await.ready(Task.unit.delayExecution(2.second).runToFuture, 3.second)
}
```
What is the difference with `.executeAsync` and without it? 
Output in the first case:
```
Task was canceled
heavy computation done
```
In the second one:
```
heavy computation done
```

**Comment by Avasil** [comment]
`runToFuture` runs synchronously on the current thread until the first asynchronous boundary so `cancelable.cancel` was called after the task has already finished

**Reply by kpodsiad** [reply]
Yeah, after I added `println(Thread.currentThread().getName)` it's clear that Task is running on `main` thread and not on the one of the given pool, thank you for explaining!

However, that behavior surprises me. According to `Task.runToFuture` docs it "Triggers the asynchronous execution" and both `Eval` and `Async` don't overwrite `Task.runToFutureOpt`, which is basically an implementation of aforementioned. I think that current description is a little bit misleading and what you wrote, `runs synchronously on the current thread until the first asynchronous boundary`, would be much more adequate.
